tc/sgx/trusted_worker_manager/enclave/work_order_processor_kme.cpp

```diff
@@ -0,0 +1,152 @@
+        EnclaveData* enclave_data) {
+
+        std::vector<tcf::WorkOrderData> in_wo_data;
+        std::vector<tcf::WorkOrderData> out_wo_data;
+        if (data_items_in.size() > 0) {
+            // data_items_in vector is sorted based out of index
+            for (auto d : data_items_in) {
+                in_wo_data.emplace_back(d.workorder_data.index,
+                                        d.workorder_data.decrypted_data);
+            }
+
+            for (auto d : data_items_out) {
+                out_wo_data.emplace_back(d.workorder_data.index,
+                                            d.workorder_data.decrypted_data);
+            }
+
+            // Convert workload_id from hex string to string
+            ByteArray workload_bytes = HexStringToBinary(workload_id);
+            std::string workload_type(workload_bytes.begin(), workload_bytes.end());
+            WorkloadProcessor *processor;
+            // Creating workload processor for "kme" enclave type. The type of
+            // workload is checked in kme_workload_plugin in case of KME and
+            // processed accordingly.
+
+            processor = WorkloadProcessor::CreateWorkloadProcessor("kme");
+            tcf::error::ThrowIf<tcf::error::WorkloadError>(
+                (processor==nullptr) ||
+                (processor->ext_work_order_info==nullptr),
+                "Failed to initialize KME workload");
+            if (workload_type == "kme-preprocess") {
+                // KME preprocess request will embed clients work order
+                // request at index 0 and ext_work_order_data at index 1
+                // of in-data. We need to parse client request
+                // to fetch client specific keys which will be wrapped
+                // during preprocess.
+                std::string orig_wo_req = ByteArrayToStr(
+                        data_items_in[0].workorder_data.decrypted_data);
+                std::string ext_data = ByteArrayToStr(
+                    data_items_in[1].workorder_data.decrypted_data);
+                WorkOrderProcessorKME kme_wo_proc;
+                kme_wo_proc.ParseJsonInput(enclave_data, orig_wo_req);
+                ExtWorkOrderInfoKME* ext_wo_info_kme = \
+                    (ExtWorkOrderInfoKME*) processor->ext_work_order_info;
+                kme_wo_proc.PopulateExtInfoData(ext_data,
+                    ext_wo_info_kme);
+            }
+
+            processor->ProcessWorkOrder(
+                    workload_type,
+                    StrToByteArray(requester_id),
+                    StrToByteArray(worker_id),
+                    StrToByteArray(work_order_id),
+                    in_wo_data,
+                    out_wo_data);
+            return out_wo_data;
+        }
+        throw tcf::error::RuntimeError("Work order inData not found");
+    }  // WorkOrderProcessorKME::ExecuteWorkOrder
+
+    ByteArray WorkOrderProcessorKME::Process(EnclaveData* enclaveData, std::string json_str) {
```

Comment:
Isn't this function same as base class `WorkOrderProcessor::Process()`? Why do we need to override this ?

Reply:
Yes. But In the absence of Process() in the WorkOrderProcessorKME class, it would invoke Process method of base class which in turn will call all the base class methods. This will avoid invoking overriden methods of WorkOrderProcessorKME class. Hence i have to add the Process() function here.

Reply:
use `virtual` keyword in `Process()` in `work_order_processor.h`

Reply:
Done